File: packages/server/src/engine.test.ts
import { describe, it, expect } from "vitest";
import { scoreMessage, generateReply } from "./engine.js";
import { findScenario } from "./scenarios.js";

describe("scoreMessage", () => {
  it("rewards balanced, empathetic and assertive messages", () => {
    const good = scoreMessage(
      "I really appreciate how slammed you've been. I'd like us to agree on a plan so the deadlines stop slipping. Can we set a weekly check-in?",
    );
    expect(good.empathy).toBeGreaterThan(55);
    expect(good.assertiveness).toBeGreaterThan(45);
    expect(good.overall).toBeGreaterThan(60);
  });

  it("penalizes aggressive, blaming language", () => {
    const bad = scoreMessage("You always miss deadlines and it's your fault the team looks stupid.");
    expect(bad.empathy).toBeLessThan(30);
    expect(bad.tips.join(" ")).toMatch(/blaming/i);
  });

  it("flags excessive hedging", () => {
    const hedgy = scoreMessage("Sorry to bother, maybe I just kind of wanted to possibly ask something?");
    expect(hedgy.assertiveness).toBeLessThan(45);
    expect(hedgy.tips.join(" ")).toMatch(/hedg/i);
  });
});

describe("generateReply", () => {
  it("warms up the counterpart for a strong message", () => {
    const scenario = findScenario("ask-for-raise")!;
    const res = generateReply(
      scenario,
      [],
      "I appreciate you making time. I'd like to talk about a raise: I shipped 3 launches and cut latency by 40%. What's possible?",
    );
    expect(res.reply).toBeTruthy();
    expect(res.mood).toMatch(/receptive/);
    expect(res.feedback.overall).toBeGreaterThan(60);
  });

  it("makes the counterpart guarded for a weak message", () => {
    const scenario = findScenario("give-feedback")!;
    const res = generateReply(scenario, [], "you never do your work");
    expect(res.mood).toMatch(/defensive/);
  });
});

File: shared/src/commonTest/kotlin/com/hardtalkai/shared/EngineTest.kt
package com.hardtalkai.shared

import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EngineTest {
    @Test
    fun rewardsBalancedEmpatheticAssertiveMessages() {
        val good = scoreMessage(
            "I really appreciate how slammed you've been. I'd like us to agree on a plan " +
                "so the deadlines stop slipping. Can we set a weekly check-in?",
        )
        assertTrue(good.empathy > 55, "empathy was ${good.empathy}")
        assertTrue(good.assertiveness > 45, "assertiveness was ${good.assertiveness}")
        assertTrue(good.overall > 60, "overall was ${good.overall}")
    }

    @Test
    fun penalizesAggressiveBlamingLanguage() {
        val bad = scoreMessage("You always miss deadlines and it's your fault the team looks stupid.")
        assertTrue(bad.empathy < 30, "empathy was ${bad.empathy}")
        assertTrue(bad.tips.any { it.contains("blaming", ignoreCase = true) }, "tips: ${bad.tips}")
    }

    @Test
    fun flagsExcessiveHedging() {
        val hedgy = scoreMessage("Sorry to bother, maybe I just kind of wanted to possibly ask something?")
        assertTrue(hedgy.assertiveness < 45, "assertiveness was ${hedgy.assertiveness}")
        assertTrue(hedgy.tips.any { it.contains("hedg", ignoreCase = true) }, "tips: ${hedgy.tips}")
    }

    @Test
    fun warmsUpCounterpartForStrongMessage() {
        val scenario = findScenario("ask-for-raise")
        assertNotNull(scenario)
        val res = generateReply(
            scenario,
            emptyList(),
            "I appreciate you making time. I'd like to talk about a raise: I shipped 3 launches " +
                "and cut latency by 40%. What's possible?",
        )
        assertTrue(res.reply.isNotEmpty())
        assertContains(res.mood, "receptive")
        assertTrue(res.feedback.overall > 60, "overall was ${res.feedback.overall}")
    }

    @Test
    fun makesCounterpartGuardedForWeakMessage() {
        val scenario = findScenario("give-feedback")
        assertNotNull(scenario)
        val res = generateReply(scenario, emptyList(), "you never do your work")
        assertContains(res.mood, "defensive")
    }
}
